### PR TITLE
Sign payload during build

### PR DESCRIPTION
**Pull request by flimzy**

### CODE REVIEW

File: internal/testdata/TestBuild_merge_YAML [hunk 1]
```diff
@@ -128,83 +82,130 @@
...
+    "head": {
+        "dig": {
+            "alg": "sha256",
+            "val": "7f90caa1a46526ab35c8c80a03fbf9cbdc97a0b6aa9d286dcb82143fe23bbb73"
+        },
+        "uuid": "9d8eafd5-77be-11ec-b485-5405db9a3e49"
     },
     "sigs": [
-        "eyJhbGciOiJFUzI1NiIsImtpZCI6ImE2YzM5MjBkLThlMzMtNDE1OS1iMzM3LTllNzQ2MTcxNmRmMSJ9.eyJ1dWlkIjoiOWQ4ZWFmZDUtNzdiZS0xMWVjLWI0ODUtNTQwNWRiOWEzZTQ5IiwidHlwIjoiYmlsbC5JbnZvaWNlIiwicmduIjoiRVMiLCJkaWciOnsiYWxnIjoic2hhMjU2IiwidmFsIjoiZGNlM2JjM2M4YmYyOGYzZDIwOWY3ODM5MTdiMzA4MmRkYzAzMzlhNjZlOWJhM2FhNjM4NDllNDM1N2RiMTQyMiJ9fQ.CbBRUWrP1ELBmsoF9KnMWB5D6f1rNkai4lMZYx83vP2oKt3XdIMv-kvME4voFWN7XDjmsFb8dnt1sD00Oi54rQ"
+        "eyJhbGciOiJFUzI1NiIsImtpZCI6ImE2YzM5MjBkLThlMzMtNDE1OS1iMzM3LTllNzQ2MTcxNmRmMSJ9.eyJ1dWlkIjoiOWQ4ZWFmZDUtNzdiZS0xMWVjLWI0ODUtNTQwNWRiOWEzZTQ5IiwidHlwIjoiYmlsbC5JbnZvaWNlIiwicmduIjoiRVMiLCJkaWciOnsiYWxnIjoic2hhMjU2IiwidmFsIjoiZGNlM2JjM2M4YmYyOGYzZDIwOWY3ODM5MTdiMzA4MmRkYzAzMzlhNjZlOWJhM2FhNjM4NDllNDM1N2RiMTQyMiJ9fQ.CbBRUWrP1ELBmsoF9KnMWB5D6f1rNkai4lMZYx83vP2oKt3XdIMv-kvME4voFWN7XDjmsFb8dnt1sD00Oi54rQ",
```
Comment (samlown) [inline internal/testdata/TestBuild_merge_YAML:208]: I don't think this is correct, there shouldn't be two signatures from build. Comparing them both on jwt.io indicates the Digest values do not match. I'd look at the source of these as I do think it's important we're testing against valid data.

File: cmd/gobl/serve.go [hunk 2]
```diff
@@ -88,8 +88,9 @@ func (s *serveOpts) version() echo.HandlerFunc {
 }
 
 type buildRequest struct {
-	Template json.RawMessage `json:"template"`
-	Data     json.RawMessage `json:"data"`
+	Template   json.RawMessage  `json:"template"`
+	Data       json.RawMessage  `json:"data"`
+	PrivateKey *dsig.PrivateKey `json:"privatekey"`
```
Comment (samlown) [inline cmd/gobl/serve.go:93]: I think I'd just set this to `Key` with an extra validation to ensure it's a private key.

**Review by samlown — APPROVED** [review]
Looks okay. I'd just look at the dual-sigs before committing, as that could be confusing.

**Comment by flimzy** [comment]
> Looks okay. I'd just look at the dual-sigs before committing, as that could be confusing.

What is the correct behavior here? Should `Sign()` remove old sigs before adding a new one?

File: cmd/gobl/serve.go (hunk 2 above)
Comment (flimzy) [reply cmd/gobl/serve.go:93]: My only reason for calling it `privatekey` here was that with the Verify method, we'll need a `publickey`, and I wanted to make it clear that they're different keys.

WDYT?

Comment (samlown) [reply cmd/gobl/serve.go:93]: It is an API, so writing more is less of a concern. Let's leave it as `privatekey` then.

**Comment by samlown** [comment]
> What is the correct behavior here? Should `Sign()` remove old sigs before adding a new one?

I think it'd be best to invalidate the request and return an error as there shouldn't be any invalid signatures there. Most likely in these examples the sigs have been copied over from elsewhere.

**Comment by flimzy** [comment]
> > What is the correct behavior here? Should `Sign()` remove old sigs before adding a new one?
> 
> I think it'd be best to invalidate the request and return an error as there shouldn't be any invalid signatures there. Most likely in these examples the sigs have been copied over from elsewhere.

Makes sense. I'll address that in a follow-up PR then.

**Pull request merged**